### DISCUSSION TITLE
Serving static assets with extensions

### DISCUSSION THREAD

**Discussion by AustinPhillipTaylor** [directus/directus#8354] [Feature Requests]
There are plenty of use-cases where a module will need to serve assets, i.e. images, text documents, etc. 

This was briefly addressed in #7488, however, I feel like we're in need of a more long-term, scalable solution. 

I like the idea of having a `public` folder adjacent to the compiled `index.js` file: 
```
 extensions/
 └── modules/
     └── my-extension/
         ├── public/
         │   └── logo.png
         └── index.js
```
and then using something along the lines of `app.use( '/my_extension_id/public', express.static( 'public' ) )` to simply expose the assets. This way, in our extensions we can point directly to the endpoint for our sources, i.e. 
```
<img src="localhost:8055/my_extension_id/public/logo.png">
```

I can't think of any reason it shouldn't be possible, thoughts?

**Comment by rijkvanzanten** [comment]
> I can't think of any reason it shouldn't be possible, thoughts?
 
As a matter of fact, I think this is already possible as is 🤔

**Reply by AustinPhillipTaylor** [reply]
Is there already an endpoint? Or are you just saying the pieces are all there?

My impression is that if there isn't already an endpoint for it, it'll probably be one of the simplest features to implement. Very little work, for a lot of benefit. 😂

**Reply by rijkvanzanten** [reply]
Your extension can pull in express from the dependency, and seeing the custom endpoint registration function exposes the router instance, there shouldn't be anything stopping you from adding express.static as a middleware currently 🤔 

(Given, I've never tried this, but like you mentioned all the pieces should be there already)

**Reply by AustinPhillipTaylor** [reply]
Interesting thought, I'll give it a shot when I get a chance.

**Comment by AustinPhillipTaylor** [comment]
@rijkvanzanten Sorry it took me so long to get back about this. Had a chance to try out the method you suggested. In short, it does work. However, I have some thoughts. 

Creating a static directory like this is admittedly very simple, though it feels a bit strange to create an entirely separate extension just to serve static files. Aside from that, it is nice that using an endpoint for this purpose makes everything just work, in very few lines of code. Since the endpoints are served directly from node, any package (like express) that is already used by the server can be easily required in a CommonJS style, which is very convenient.  

I do find myself wondering if it'd be easily possible to just create a global public endpoint, whereas a sub-path is automatically created to serve an adjacent 'public' folder whenever a module is registered (similar to what I described in my original post). It's not terribly important, just seems a bit redundant if each module author creates a custom endpoint simply for the sake of displaying a couple images (logos, maybe icons) in their module. 

On the plus side, if it is left the way it is, it's extremely light weight. Minified, the endpoint is only 239 characters (plus whatever assets will be served). So a hundred different modules could use this method and it wouldn't add much at all to the load time.

**Reply by rijkvanzanten** [reply]
> I do find myself wondering if it'd be easily possible to just create a global public endpoint, whereas a sub-path is automatically created to serve an adjacent 'public' folder whenever a module is registered (similar to what I described in my original post). It's not terribly important, just seems a bit redundant if each module author creates a custom endpoint simply for the sake of displaying a couple images (logos, maybe icons) in their module.

Yeah, I get what you're saying, but I'm not sure if this particular use case applies to 80%+ of custom extensions 🤔 I would also want to prevent people from accidentally making stuff public that shouldn't be public

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
